kernel/src/process.rs

```diff
@@ -933,11 +933,32 @@ impl<'a> Process<'a> {
             if let Some(load_result) =
                 load(tbf_header, remaining_app_memory) {
 
+                // First determine how much space we need in the application's
+                // memory space just for kernel and grant state. We need to make
+                // sure we allocate enough memory just for that.
+
+                // Make room for grant pointers.
+                let grant_ptr_size = mem::size_of::<*const usize>();
+                let grant_ptrs_num = read_volatile(&grant::CONTAINER_COUNTER);
+                let grant_ptrs_offset = grant_ptrs_num * grant_ptr_size;
+
+                // Allocate memory for callback ring buffer.
+                let callback_size = mem::size_of::<Task>();
+                let callback_len = 10;
+                let callbacks_offset = callback_len * callback_size;
+
+                // Need to make sure that the amount of memory we allocate for
+                // this process at least covers this state.
+                if min_app_ram_size < (grant_ptrs_offset + callbacks_offset) as u32 {
```

Comment:
I'm a little wary that kernel_memory_break is computed separately from min_app_ram_size. I.e., as written, the amount kernel_memory_break is moved should be equal to min_app_ram_size, but it could be different, e.g., if some other structure were added but forgotten to be added to min_app_ram_size. Is it possible to just define min_app_ram_size to be the amount that kernel_memory_break has moved?

Reply:
I see what you mean and it's a good point. I'm not sure how to implement it, exactly, as the kernel_memory_break pointer is set based on already having the allocated memory (rounded up to a power of two).

Reply:
Fair enough -- I don't think this is a concrete issue, just something that requires extra care from someone who modifies this code in the future.